File: packages/firebase_app_check/firebase_app_check/android/src/main/java/io/flutter/plugins/firebase/appcheck/FlutterFirebaseAppCheckPlugin.java
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebase.appcheck;

import static io.flutter.plugins.firebase.core.FlutterFirebasePluginRegistry.registerPlugin;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.TaskCompletionSource;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.FirebaseApp;
import com.google.firebase.appcheck.AppCheckTokenResult;
import com.google.firebase.appcheck.FirebaseAppCheck;
import com.google.firebase.appcheck.debug.DebugAppCheckProviderFactory;
import com.google.firebase.appcheck.playintegrity.PlayIntegrityAppCheckProviderFactory;
import com.google.firebase.appcheck.safetynet.SafetyNetAppCheckProviderFactory;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugins.firebase.core.FlutterFirebasePlugin;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class FlutterFirebaseAppCheckPlugin
    implements FlutterFirebasePlugin, FlutterPlugin, MethodCallHandler {

  private static final String METHOD_CHANNEL_NAME = "plugins.flutter.io/firebase_app_check";
  private final Map<EventChannel, TokenChannelStreamHandler> streamHandlers = new HashMap<>();
  private final String TAG = "FLTAppCheckPlugin";

  private final String debugProvider = "debug";
  private final String safetyNetProvider = "safetyNet";
  private final String playIntegrity = "playIntegrity";

  @Nullable private BinaryMessenger messenger;

  private MethodChannel channel;

  private void initInstance(BinaryMessenger messenger) {
    registerPlugin(METHOD_CHANNEL_NAME, this);
    channel = new MethodChannel(messenger, METHOD_CHANNEL_NAME);
    channel.setMethodCallHandler(this);

    this.messenger = messenger;
  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    initInstance(binding.getBinaryMessenger());
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
    channel = null;
    messenger = null;

    removeEventListeners();
  }

  private FirebaseAppCheck getAppCheck(Map<String, Object> arguments) {
    String appName = (String) Objects.requireNonNull(arguments.get("appName"));
    FirebaseApp app = FirebaseApp.getInstance(appName);
    return FirebaseAppCheck.getInstance(app);
  }

  private Map<String, Object> tokenResultToMap(AppCheckTokenResult result) {
    Map<String, Object> output = new HashMap<>();
    output.put("token", result.getToken());
    return output;
  }

  private Task<Void> activate(Map<String, Object> arguments) {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            String provider = (String) Objects.requireNonNull(arguments.get("androidProvider"));

            switch (provider) {
              case debugProvider:
                {
                  FirebaseAppCheck firebaseAppCheck = FirebaseAppCheck.getInstance();
                  firebaseAppCheck.installAppCheckProviderFactory(
                      DebugAppCheckProviderFactory.getInstance());
                  break;
                }
              case safetyNetProvider:
                {
                  FirebaseAppCheck firebaseAppCheck = getAppCheck(arguments);
                  firebaseAppCheck.installAppCheckProviderFactory(
                      SafetyNetAppCheckProviderFactory.getInstance());
                  break;
                }
              case playIntegrity:
                {
                  FirebaseAppCheck firebaseAppCheck = getAppCheck(arguments);
                  firebaseAppCheck.installAppCheckProviderFactory(
                      PlayIntegrityAppCheckProviderFactory.getInstance());
                  break;
                }
            }
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<Map<String, Object>> getToken(Map<String, Object> arguments) {
    TaskCompletionSource<Map<String, Object>> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            FirebaseAppCheck firebaseAppCheck = getAppCheck(arguments);
            Boolean forceRefresh = (Boolean) Objects.requireNonNull(arguments.get("forceRefresh"));
            AppCheckTokenResult tokenResult = Tasks.await(firebaseAppCheck.getToken(forceRefresh));

            taskCompletionSource.setResult(tokenResultToMap(tokenResult));
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<Void> setTokenAutoRefreshEnabled(Map<String, Object> arguments) {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            FirebaseAppCheck firebaseAppCheck = getAppCheck(arguments);
            Boolean isTokenAutoRefreshEnabled =
                (Boolean) Objects.requireNonNull(arguments.get("isTokenAutoRefreshEnabled"));
            firebaseAppCheck.setTokenAutoRefreshEnabled(isTokenAutoRefreshEnabled);

            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private Task<String> registerTokenListener(Map<String, Object> arguments) {
    TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            String appName = (String) Objects.requireNonNull(arguments.get("appName"));
            FirebaseAppCheck firebaseAppCheck = getAppCheck(arguments);

            final TokenChannelStreamHandler handler =
                new TokenChannelStreamHandler(firebaseAppCheck);
            final String name = METHOD_CHANNEL_NAME + "/token/" + appName;
            final EventChannel channel = new EventChannel(messenger, name);
            channel.setStreamHandler(handler);
            streamHandlers.put(channel, handler);

            taskCompletionSource.setResult(name);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  @Override
  public void onMethodCall(MethodCall call, @NonNull final Result result) {
    Task<?> methodCallTask;

    switch (call.method) {
      case "FirebaseAppCheck#activate":
        methodCallTask = activate(call.arguments());
        break;
      case "FirebaseAppCheck#getToken":
        methodCallTask = getToken(call.arguments());
        break;
      case "FirebaseAppCheck#setTokenAutoRefreshEnabled":
        methodCallTask = setTokenAutoRefreshEnabled(call.arguments());
        break;
      case "FirebaseAppCheck#registerTokenListener":
        methodCallTask = registerTokenListener(call.arguments());
        break;
      default:
        result.notImplemented();
        return;
    }

    methodCallTask.addOnCompleteListener(
        task -> {
          if (task.isSuccessful()) {
            result.success(task.getResult());
          } else {
            Exception exception = task.getException();
            result.error(
                "firebase_app_check",
                exception != null ? exception.getMessage() : null,
                getExceptionDetails(exception));
          }
        });
  }

  private Map<String, Object> getExceptionDetails(@Nullable Exception exception) {
    Map<String, Object> details = new HashMap<>();
    details.put("code", "unknown");
    if (exception != null) {
      details.put("message", exception.getMessage());
    } else {
      details.put("message", "An unknown error has occurred.");
    }
    return details;
  }

  @Override
  public Task<Map<String, Object>> getPluginConstantsForFirebaseApp(FirebaseApp firebaseApp) {
    TaskCompletionSource<Map<String, Object>> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  @Override
  public Task<Void> didReinitializeFirebaseCore() {
    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();

    cachedThreadPool.execute(
        () -> {
          try {
            taskCompletionSource.setResult(null);
          } catch (Exception e) {
            taskCompletionSource.setException(e);
          }
        });

    return taskCompletionSource.getTask();
  }

  private void removeEventListeners() {
    for (EventChannel eventChannel : streamHandlers.keySet()) {
      EventChannel.StreamHandler streamHandler = streamHandlers.get(eventChannel);
      streamHandler.onCancel(null);
      eventChannel.setStreamHandler(null);
    }
    streamHandlers.clear();
  }
}

File: packages/firebase_app_check/firebase_app_check/android/src/main/kotlin/io/flutter/plugins/firebase/appcheck/FlutterFirebaseAppCheckPlugin.kt
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebase.appcheck

import androidx.annotation.NonNull
import androidx.annotation.Nullable
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.TaskCompletionSource
import com.google.android.gms.tasks.Tasks
import com.google.firebase.FirebaseApp
import com.google.firebase.appcheck.AppCheckToken
import com.google.firebase.appcheck.FirebaseAppCheck
import com.google.firebase.appcheck.debug.DebugAppCheckProviderFactory
import com.google.firebase.appcheck.playintegrity.PlayIntegrityAppCheckProviderFactory
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugins.firebase.core.FlutterFirebasePlugin
import io.flutter.plugins.firebase.core.FlutterFirebasePlugin.cachedThreadPool
import io.flutter.plugins.firebase.core.FlutterFirebasePluginRegistry.registerPlugin
import io.flutter.plugins.firebase.firebaseappcheck.AppCheckWebProvider
import io.flutter.plugins.firebase.firebaseappcheck.AppCheckAndroidProvider
import io.flutter.plugins.firebase.firebaseappcheck.AppCheckAppleProvider
import io.flutter.plugins.firebase.firebaseappcheck.FirebaseAppCheckHostApi

class FlutterFirebaseAppCheckPlugin : FlutterFirebasePlugin, FlutterPlugin, MethodCallHandler, FirebaseAppCheckHostApi {

    companion object {
        private const val METHOD_CHANNEL_NAME = "plugins.flutter.io/firebase_app_check"
        private const val DEBUG_PROVIDER = "debug"
        private const val PLAY_INTEGRITY = "playIntegrity"
    }

    private val streamHandlers = mutableMapOf<EventChannel, TokenChannelStreamHandler>()

    @Nullable
    private var messenger: BinaryMessenger? = null

    private var channel: MethodChannel? = null

    private fun initInstance(messenger: BinaryMessenger) {
        registerPlugin(METHOD_CHANNEL_NAME, this)
        channel = MethodChannel(messenger, METHOD_CHANNEL_NAME)
        channel?.setMethodCallHandler(this)

        // Set up Pigeon API
        FirebaseAppCheckHostApi.setUp(messenger, this)

        this.messenger = messenger
    }

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        initInstance(binding.binaryMessenger)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel = null
        
        // Clean up Pigeon API
        FirebaseAppCheckHostApi.setUp(binding.binaryMessenger, null)
        
        messenger = null

        removeEventListeners()
    }

    private fun getAppCheck(arguments: Map<String, Any>): FirebaseAppCheck {
        val appName = arguments["appName"] as String
        val app = FirebaseApp.getInstance(appName)
        return FirebaseAppCheck.getInstance(app)
    }

    private fun getLimitedUseAppCheckToken(arguments: Map<String, Any>): Task<String> {
        val taskCompletionSource = TaskCompletionSource<String>()

        cachedThreadPool.execute {
            try {
                val firebaseAppCheck = getAppCheck(arguments)
                val tokenResult = Tasks.await(firebaseAppCheck.limitedUseAppCheckToken)
                taskCompletionSource.setResult(tokenResult.token)
            } catch (e: Exception) {
                taskCompletionSource.setException(e)
            }
        }

        return taskCompletionSource.task
    }

    private fun activate(arguments: Map<String, Any>): Task<Void?> {
        val taskCompletionSource = TaskCompletionSource<Void?>()

        cachedThreadPool.execute {
            try {
                val provider = arguments["androidProvider"] as String

                when (provider) {
                    DEBUG_PROVIDER -> {
                        val firebaseAppCheck = getAppCheck(arguments)
                        firebaseAppCheck.installAppCheckProviderFactory(
                            DebugAppCheckProviderFactory.getInstance()
                        )
                    }
                    PLAY_INTEGRITY -> {
                        val firebaseAppCheck = getAppCheck(arguments)
                        firebaseAppCheck.installAppCheckProviderFactory(
                            PlayIntegrityAppCheckProviderFactory.getInstance()
                        )
                    }
                }
                taskCompletionSource.setResult(null)
            } catch (e: Exception) {
                taskCompletionSource.setException(e)
            }
        }

        return taskCompletionSource.task
    }

    private fun getToken(arguments: Map<String, Any>): Task<String> {
        val taskCompletionSource = TaskCompletionSource<String>()

        cachedThreadPool.execute {
            try {
                val firebaseAppCheck = getAppCheck(arguments)
                val forceRefresh = arguments["forceRefresh"] as Boolean
                val tokenResult = Tasks.await(firebaseAppCheck.getAppCheckToken(forceRefresh))

                taskCompletionSource.setResult(tokenResult.token)
            } catch (e: Exception) {
                taskCompletionSource.setException(e)
            }
        }

        return taskCompletionSource.task
    }

    private fun setTokenAutoRefreshEnabled(arguments: Map<String, Any>): Task<Void?> {
        val taskCompletionSource = TaskCompletionSource<Void?>()

        cachedThreadPool.execute {
            try {
                val firebaseAppCheck = getAppCheck(arguments)
                val isTokenAutoRefreshEnabled = arguments["isTokenAutoRefreshEnabled"] as Boolean
                firebaseAppCheck.setTokenAutoRefreshEnabled(isTokenAutoRefreshEnabled)

                taskCompletionSource.setResult(null)
            } catch (e: Exception) {
                taskCompletionSource.setException(e)
            }
        }

        return taskCompletionSource.task
    }

    private fun registerTokenListener(arguments: Map<String, Any>): Task<String> {
        val taskCompletionSource = TaskCompletionSource<String>()

        cachedThreadPool.execute {
            try {
                val appName = arguments["appName"] as String
                val firebaseAppCheck = getAppCheck(arguments)

                val handler = TokenChannelStreamHandler(firebaseAppCheck)
                val name = "$METHOD_CHANNEL_NAME/token/$appName"
                val eventChannel = EventChannel(messenger, name)
                eventChannel.setStreamHandler(handler)
                streamHandlers[eventChannel] = handler

                taskCompletionSource.setResult(name)
            } catch (e: Exception) {
                taskCompletionSource.setException(e)
            }
        }

        return taskCompletionSource.task
    }

    override fun onMethodCall(call: MethodCall, @NonNull result: Result) {
        val arguments = call.arguments<Map<String, Any>>() ?: return
        val methodCallTask: Task<*> = when (call.method) {
            "FirebaseAppCheck#activate" -> activate(arguments)
            "FirebaseAppCheck#getToken" -> getToken(arguments)
            "FirebaseAppCheck#setTokenAutoRefreshEnabled" -> setTokenAutoRefreshEnabled(arguments)
            "FirebaseAppCheck#registerTokenListener" -> registerTokenListener(arguments)
            "FirebaseAppCheck#getLimitedUseAppCheckToken" -> getLimitedUseAppCheckToken(arguments)
            else -> {
                result.notImplemented()
                return
            }
        }

        methodCallTask.addOnCompleteListener { task ->
            if (task.isSuccessful) {
                result.success(task.result)
            } else {
                val exception = task.exception
                result.error(
                    "firebase_app_check",
                    exception?.message,
                    getExceptionDetails(exception)
                )
            }
        }
    }

    private fun getExceptionDetails(@Nullable exception: Exception?): Map<String, Any> {
        val details = mutableMapOf<String, Any>()
        details["code"] = "unknown"
        if (exception != null) {
            details["message"] = exception.message ?: "An unknown error has occurred."
        } else {
            details["message"] = "An unknown error has occurred."
        }
        return details
    }

    override fun getPluginConstantsForFirebaseApp(firebaseApp: FirebaseApp): Task<Map<String, Any>?> {
        val taskCompletionSource = TaskCompletionSource<Map<String, Any>?>()

        cachedThreadPool.execute {
            try {
                taskCompletionSource.setResult(null)
            } catch (e: Exception) {
                taskCompletionSource.setException(e)
            }
        }

        return taskCompletionSource.task
    }

    override fun didReinitializeFirebaseCore(): Task<Void?> {
        val taskCompletionSource = TaskCompletionSource<Void?>()

        cachedThreadPool.execute {
            try {
                taskCompletionSource.setResult(null)
            } catch (e: Exception) {
                taskCompletionSource.setException(e)
            }
        }

        return taskCompletionSource.task
    }

    private fun removeEventListeners() {
        for ((eventChannel, streamHandler) in streamHandlers) {
            streamHandler.onCancel(null)
            eventChannel.setStreamHandler(null)
        }
        streamHandlers.clear()
    }

    // Pigeon API implementation
    override fun activate(
        appName: String,
        webProvider: AppCheckWebProvider,
        androidProvider: AppCheckAndroidProvider,
        appleProvider: AppCheckAppleProvider,
        callback: (kotlin.Result<Unit>) -> Unit
    ) {
        cachedThreadPool.execute {
            try {
                val app = FirebaseApp.getInstance(appName)
                val firebaseAppCheck = FirebaseAppCheck.getInstance(app)

                when (androidProvider.providerName) {
                    DEBUG_PROVIDER -> {
                        firebaseAppCheck.installAppCheckProviderFactory(
                            DebugAppCheckProviderFactory.getInstance()
                        )
                    }
                    PLAY_INTEGRITY -> {
                        firebaseAppCheck.installAppCheckProviderFactory(
                            PlayIntegrityAppCheckProviderFactory.getInstance()
                        )
                    }
                }
                callback(kotlin.Result.success(Unit))
            } catch (e: Exception) {
                callback(kotlin.Result.failure(e))
            }
        }
    }

    override fun getToken(appName: String, forceRefresh: Boolean, callback: (kotlin.Result<String?>) -> Unit) {
        cachedThreadPool.execute {
            try {
                val app = FirebaseApp.getInstance(appName)
                val firebaseAppCheck = FirebaseAppCheck.getInstance(app)
                val tokenResult = Tasks.await(firebaseAppCheck.getAppCheckToken(forceRefresh))
                callback(kotlin.Result.success(tokenResult.token))
            } catch (e: Exception) {
                callback(kotlin.Result.failure(e))
            }
        }
    }

    override fun setTokenAutoRefreshEnabled(
        appName: String,
        isTokenAutoRefreshEnabled: Boolean,
        callback: (kotlin.Result<Unit>) -> Unit
    ) {
        cachedThreadPool.execute {
            try {
                val app = FirebaseApp.getInstance(appName)
                val firebaseAppCheck = FirebaseAppCheck.getInstance(app)
                firebaseAppCheck.setTokenAutoRefreshEnabled(isTokenAutoRefreshEnabled)
                callback(kotlin.Result.success(Unit))
            } catch (e: Exception) {
                callback(kotlin.Result.failure(e))
            }
        }
    }

    override fun getLimitedUseToken(appName: String, callback: (kotlin.Result<String>) -> Unit) {
        cachedThreadPool.execute {
            try {
                val app = FirebaseApp.getInstance(appName)
                val firebaseAppCheck = FirebaseAppCheck.getInstance(app)
                val tokenResult = Tasks.await(firebaseAppCheck.limitedUseAppCheckToken)
                callback(kotlin.Result.success(tokenResult.token))
            } catch (e: Exception) {
                callback(kotlin.Result.failure(e))
            }
        }
    }
} 